core/commands/storage/upload/upload/dc_repair_router.go

```diff
@@ -202,44 +212,338 @@ returns the repairer's signed contract to the invoker.`,
+			log.Errorf("repair error: %s", err.Error())
 		}
 	}()
 }
 
+func submitSignedRepairContract(ctxParams *uh.ContextParams, params *RepairContractParams) (*guardpb.RepairContractResponse, error) {
+
+	repairReq := &guardpb.RepairContract{
+		FileHash:             params.FileHash,
+		FileSize:             params.FileSize,
+		RepairPid:            params.RepairPid,
+		LostShardHash:        params.LostShardHashes,
+		RepairSignTime:       time.Now().UTC(),
+		RepairContractId:     params.RepairContractId,
+		DownloadContractId:   params.DownloadContractId,
+		RepairRewardAmount:   params.RepairRewardAmount,
+		DownloadRewardAmount: params.DownloadRewardAmount,
+	}
+	sig, err := crypto.Sign(ctxParams.N.PrivateKey, repairReq)
+	if err != nil {
+		return nil, err
+	}
+	repairReq.RepairSignature = sig
+	var repairResp *guardpb.RepairContractResponse
+	err = grpc.GuardClient(ctxParams.Cfg.Services.GuardDomain).WithContext(ctxParams.Ctx, func(ctx context.Context,
+		client guardpb.GuardServiceClient) error {
+		repairResp, err = client.SubmitRepairContract(ctx, repairReq)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return repairResp, nil
+}
+
+func uploadShards(ctxParams *uh.ContextParams, repairContract *guardpb.RepairContract) (*guardpb.FileStoreStatus, error) {
+	repairContractReq := &guardpb.RequestRepairContracts{
+		FileHash:       repairContract.FileHash,
+		RepairNode:     repairContract.RepairPid,
+		RepairSignTime: repairContract.RepairSignTime,
+	}
+	sig, err := crypto.Sign(ctxParams.N.PrivateKey, repairContractReq)
+	if err != nil {
+		return nil, err
+	}
+	repairContractReq.RepairSignature = sig
+	var statusMeta *guardpb.FileStoreStatus
+	err = grpc.GuardClient(ctxParams.Cfg.Services.GuardDomain).WithContext(ctxParams.Ctx, func(ctx context.Context,
+		client guardpb.GuardServiceClient) error {
+		repairContractResp, err := client.RequestForRepairContracts(ctx, repairContractReq)
+		if err != nil {
+			return err
+		}
+		statusMeta = repairContractResp.Status
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	contracts := statusMeta.Contracts
+	if len(contracts) <= 0 {
+		return nil, errors.New("length of contracts is 0")
+	}
+	ssId, _ := uh.SplitContractId(contracts[0].ContractId)
+	shardIndexes := make([]int, 0)
+	shardHashes := make([]string, 0)
+	contractMap := map[string]*guardpb.Contract{}
+	for _, contract := range contracts {
+		if contract.State == guardpb.Contract_LOST {
+			contractMap[contract.ShardHash] = contract
+			shardHashes = append(shardHashes, contract.ShardHash)
+			shardIndexes = append(shardIndexes, int(contract.ShardIndex))
+		}
+	}
+	rss, err := sessions.GetRenterSession(ctxParams, ssId, repairContract.FileHash, shardHashes)
+	if err != nil {
+		return nil, err
+	}
+	hp := uh.GetHostsProvider(ctxParams, make([]string, 0))
+	m := contracts[0].ContractMeta
+	for index, shardHash := range rss.ShardHashes {
+		go func(i int, h string) {
```

Comment:
this part of code mostly should be same some existing, the only difference is target status, can we try to put it in a function to make it DRY?

Reply:
I considered to refine few repeated lines into new functions, but it could barely improve the code, because the functions are too small, otherwise refactoring the upload_shard.go to add parameters of new flag and repair contracts and filter lots of logic, actually, this part of logic is not quite same with upload function, so I keep code here.

Reply:
The thing is this part of code is very sensitive,  should be core of our code, it is better to manage them together. Anyway it is not big issue, let us quick discuss it today.